### DISCUSSION TITLE
Using api as dependency configuration confused me

### DISCUSSION THREAD

**Discussion by JackEblan** [android/nowinandroid#1393] [Q&A]
Why use api in [core/data/build.gradle.kts](https://github.com/android/nowinandroid/blob/7980c1d0dfde13d0f4c6ece581589b1ee45a501c/core/data/build.gradle.kts)?

If I understood correctly, feature modules can see the dependency of the core:data module like core:network, core:datastore, core:database. I think the interfaces from data layer is what they needed. There should be an explanation to this. Thanks.

**Comment by Jaehwa-Noh** [comment]
I'm not sure I am fully understand your question.
If you want to know what different between `api` and `implementation`, you can see this [documentation](https://developer.android.com/build/dependencies#dependency_configurations)

**Reply by JackEblan** [reply]
Hello @Jaehwa-Noh, thanks for the response.

My confusion is that _:core:data_ exposes the _:core:model_ to **FEATURE Modules** but the code here https://github.com/android/nowinandroid/blob/7980c1d0dfde13d0f4c6ece581589b1ee45a501c/build-logic/convention/src/main/kotlin/AndroidFeatureConventionPlugin.kt#L42 already tells that _:core:model_ is already accessible because https://github.com/android/nowinandroid/blob/7980c1d0dfde13d0f4c6ece581589b1ee45a501c/core/ui/build.gradle.kts#L33 exposes those models
If you remove https://github.com/android/nowinandroid/blob/7980c1d0dfde13d0f4c6ece581589b1ee45a501c/build-logic/convention/src/main/kotlin/AndroidFeatureConventionPlugin.kt#L42 It would still works because _:core:data_ already exposes those models from https://github.com/android/nowinandroid/blob/7980c1d0dfde13d0f4c6ece581589b1ee45a501c/core/database/build.gradle.kts#L33 It's just confusing me on which module really tells the truth. Sometimes it duplicates the dependency. Sorry for confusing you.

Also, in https://github.com/android/nowinandroid/blob/7980c1d0dfde13d0f4c6ece581589b1ee45a501c/core/data/build.gradle.kts#L36
Does it mean that **FEATURE Modules** can actually see this class https://github.com/android/nowinandroid/blob/7980c1d0dfde13d0f4c6ece581589b1ee45a501c/core/datastore/src/main/kotlin/com/google/samples/apps/nowinandroid/core/datastore/NiaPreferencesDataSource.kt#L29

**Reply by Jaehwa-Noh** [reply]
To sum up, Feature -> (implement) Core:UI -> (api) model 
But, where did you find `api(projects.core.datastore)`?
I can't see the connection from `Feature` to `Datastore`.

**Reply by Jaehwa-Noh** [reply]
Aha, I found it.
I confused with Core:UI what you mentioned.
https://github.com/android/nowinandroid/blob/7980c1d0dfde13d0f4c6ece581589b1ee45a501c/feature/bookmarks/build.gradle.kts#L27-L28

Then the dependency like this.
Feature -> (implementation) data -> (api) datastore

As I shown you the documentation, `data` export `datastore` and `Feature` can see it.
>it's letting Gradle know that the module wants to transitively ***export that dependency to other modules***, so that it's available to them at both runtime and compile time.

And @yongsuk44 mentioned that, both were not directly connected. Look his [comment](https://github.com/android/nowinandroid/discussions/1393#discussioncomment-9210414).

**Comment by yongsuk44** [comment]
@JackEblan 

As you mentioned, it is possible to access 'network', 'datastore', and 'database' from the 'feature' module. However, you need to understand that the 'feature' module does not directly depend on these modules instead, it accesses them indirectly through interfaces implemented by each Repository in the 'data' module.

Additionally, the reason the 'data' module uses 'api' is not just because it is used by the 'feature' module, but as you can see from the `SyncWorker` in the 'sync' module, it is used elsewhere too.
